File: 20.js
/**
 * @param {string} str
 * @return {boolean}
 */
const isValid = (str) => {
    const dict = new Map([
        ["}", "{"],
        ["]", "["],
        [")", "("]
    ]);
    const stack = [];
    const strLength = str.length;
    for (let i = 0; i < strLength; i++) {
        if ([...dict.values()].indexOf(str[i]) > -1) {
            stack.push(str[i]);
        } else {
            if (stack[stack.length-1] !== dict.get(str[i])) {
                return false;
            }
            stack.pop();
        }
    }
    return !stack.length;
};

File: 20.py
# First solution
'''
Complexity:
Space: O(n)
Time: O(n)
'''
class Solution:
    def isValid(self, s: str) -> bool:
        stack = []
        map = {
            '}': '{',
            ']': '[',
            ')': '('
        }
        for symbol in s:
            if symbol in map.values():
                stack.append(symbol)
            elif symbol in map.keys():
                if stack == [] or map[symbol] != stack.pop():
                    return False
            else:
                return False
        return stack == []


# Second solution (enhanced 1st solution)
class Solution:
    def isValid(self, s: str) -> bool:
        pair = {
            "(": ")",
            "{": "}",
            "[": "]",
        }
        stack = []
        for c in s:
            if c in pair:
                stack.append(c)
            else:
                if not stack:
                    return False

                prev = stack.pop()
                if c != pair[prev]:
                    return False
        return len(stack) == 0
